File: src/rust/src/single_cell/nebula_gpu.rs
//! GPU-accelerated NEBULA.
//!
//! Mirrors `bixverse::rs_nebula_sc` argument for argument. Only stage two, the
//! per-gene penalised fits, moves to the device; cell ordering, batching, the
//! dispersion shrinkage and the Wald test are the CPU code in `bixverse-rs`.
//! Stage two runs in `f32` on the device, so the answers sit close to the CPU
//! path rather than on it.

use crate::ensure_gpu;
use crate::single_cell::sc_utils::nebula_res_to_r_list;
use bixverse_rs::gpu::sc_gpu::nebula_gpu::run_nebula_gpu;
use bixverse_rs::prelude::*;
use bixverse_rs::single_cell::sc_analysis::nebula::NebulaScParams;
use cubecl::wgpu::{WgpuDevice, WgpuRuntime};
use cubecl::Runtime;
use extendr_api::*;

/////////////
// extendr //
/////////////

extendr_module! {
    // module
    mod nebula_gpu;
    // functions
    fn rs_nebula_sc_gpu;
}

////////////
// NEBULA //
////////////

/// GPU: fit the NEBULA negative binomial gamma mixed model over single cells
///
/// @description
/// `r lifecycle::badge("experimental")`
/// GPU equivalent of `bixverse::rs_nebula_sc`. Stage two of NEBULA, the
/// per-gene penalised fits, is dispatched to the WGPU backend in `f32` and
/// finished on the host in `f64`. Everything else, including the streaming of
/// the counts out of the gene-major store in batches, the subject ordering, the
/// dispersion shrinkage and the Wald test, is the CPU code. REML is not
/// implemented on the device and is rejected.
///
/// @param f_path_genes String. Path to the `counts_genes.bin` file.
/// @param f_path_cells String. Path to the `counts_cells.bin` file. Only read
/// when `offset` is `NULL`, to take the library sizes.
/// @param cells_to_keep Integer vector. 0-indexed(!) global positions of the
/// cells to analyse, in any order. Must not hold duplicates.
/// @param gene_indices Integer vector. 0-indexed(!) positions of the genes to
/// fit.
/// @param subject_ids Integer vector. 0-indexed(!) subject label per global
/// cell. One entry per cell in the store, not per cell in `cells_to_keep`.
/// @param design Numeric matrix. Predictors of cells x coefficients, rows
/// aligned to `cells_to_keep` and including an intercept.
/// @param offset Optional numeric vector. Strictly positive scaling factor per
/// selected cell, aligned to `cells_to_keep`. `NULL` uses the library sizes.
/// @param nebula_params Named list. The NEBULA parameters, see
/// [params_nebula_gpu()], plus either `coef` (a 0-indexed(!) coefficient) or
/// `contrast` (one weight per coefficient).
/// @param verbose Integer. `0L` - quiet; `1L` - normal verbosity; `2L` -
/// detailed verbosity.
///
/// @returns A list with the following elements
/// \itemize{
///   \item gene_idx - Integer. 0-indexed positions of the genes that survived
///   NEBULA's own expression filter.
///   \item coefficients - Numeric matrix of genes x coefficients. The fixed
///   effects on the design scale.
///   \item se - Numeric matrix of genes x coefficients. The standard errors.
///   \item subject_overdispersion - Numeric. NEBULA's `sigma^2`.
///   \item cell_overdispersion - Numeric. NEBULA's `phi^-1`.
///   \item cell_overdispersion_shrunk - Numeric or `NULL`. The cell-level
///   overdispersion after empirical Bayes shrinkage, when it was requested.
///   \item convergence - Integer. NEBULA's convergence code. At or below `-20`
///   is a likely failure.
///   \item sigma_at_bound - Boolean. Whether the subject-level variance
///   finished pinned on its lower bound.
///   \item log_fc - Numeric. Effect of the tested coefficient or contrast, on
///   the natural log scale.
///   \item effect_se - Numeric. Standard error of that effect.
///   \item z - Numeric. The Wald statistic.
///   \item p_values - Numeric. Two-sided p-values.
///   \item fdr - Numeric. Benjamini-Hochberg adjusted p-values.
/// }
///
/// @references He, et al., Commun Biol, 2021
///
/// @export
///
/// @keywords internal
#[extendr]
#[allow(clippy::too_many_arguments)]
fn rs_nebula_sc_gpu(
    f_path_genes: String,
    f_path_cells: String,
    cells_to_keep: Vec<i32>,
    gene_indices: Vec<i32>,
    subject_ids: Vec<i32>,
    design: RMatrix<f64>,
    offset: Nullable<Vec<f64>>,
    nebula_params: List,
    verbose: usize,
) -> Result<List> {
    ensure_gpu()?;

    let cells_to_keep: Vec<usize> = cells_to_keep.r_int_convert();
    let gene_indices: Vec<usize> = gene_indices.r_int_convert();
    let subject_ids: Vec<usize> = subject_ids.r_int_convert();

    let n_coef = design.ncols();
    // Column-major out of R, row-major into `run_nebula_gpu`.
    let design = mat_to_flat_row_major(r_matrix_to_faer(&design));

    let params = NebulaScParams::from_r_list(nebula_params)?;

    let gene_reader = ParallelSparseReader::new(&f_path_genes).to_extendr()?;
    let cell_reader = ParallelSparseReader::new(&f_path_cells).to_extendr()?;

    let offset = match offset {
        Nullable::NotNull(o) => Some(o),
        Nullable::Null => None,
    };

    let device: WgpuDevice = Default::default();

    let res = run_nebula_gpu::<WgpuRuntime, _>(
        &gene_reader,
        &cell_reader,
        &cells_to_keep,
        &gene_indices,
        &subject_ids,
        &design,
        n_coef,
        offset.as_deref(),
        &params,
        device.clone(),
        verbose,
    )
    .to_extendr()?;

    // force VRAM memory clean up to avoid memory leaks
    let client = WgpuRuntime::client(&device);
    client.memory_cleanup();

    Ok(nebula_res_to_r_list(res))
}

File: R/nebula_gpu.R
# ------------------------------------------------------------------------------
# GPU-accelerated NEBULA:
# - Wraps `rs_nebula_sc_gpu`.
# - Mirrors `bixverse::nebula_sc()` and reuses its design, tested-coefficient
#   and result helpers, so the `ScNebula` result is the same class with the
#   same columns.
# - bixverse dispatches on its internal `ScOrScSubset` union. That union is not
#   exported, so the two exported classes get a method each and both delegate to
#   `.nebula_gpu()`.
# ------------------------------------------------------------------------------

# nebula (gpu) -----------------------------------------------------------------

#' Run NEBULA on single cells on the GPU
#'
#' @description
#' GPU counterpart of [bixverse::nebula_sc()]. Stage two of NEBULA, the
#' per-gene penalised fits, runs on the WGPU backend in `f32` and is finished on
#' the host in `f64`. Cell ordering, gene batching, the dispersion shrinkage and
#' the Wald test are the CPU code. Expect estimates close to the CPU ones, not
#' identical to them.
#'
#' REML is not implemented on the device, so [params_nebula_gpu()] does not
#' carry it. Everything else, the design handling, the result class and the
#' downstream code, is identical to the CPU version.
#'
#' @param object `SingleCells` or `SingleCellsSubset` class from `bixverse`.
#' @param subject_col String. The column in the obs table holding the subject
#' (donor) identifier. This is what the random effect is over.
#' @param design Formula. The experimental design, evaluated against the obs
#' table, e.g. `~ condition` or `~ condition + age`. Include the intercept.
#' @param coef Optional integer or character. Which coefficient of the design
#' the Wald test reports, as a 1-based column position or a column name.
#' Defaults to the last column.
#' @param contrast Optional numeric vector. One weight per design column.
#' Mutually exclusive with `coef`.
#' @param genes_to_use Optional character vector. The genes to fit. Defaults to
#' every gene in the object, which is usually too many.
#' @param offset Optional numeric vector. Strictly positive scaling factor per
#' cell, aligned to the cells that survive the design. Defaults to `NULL`,
#' which uses the library sizes.
#' @param nebula_params A list, see [params_nebula_gpu()]. The list has the
#' following parameters:
#' \itemize{
#'   \item nebula_method - String. One of `c("ln", "hl")`.
#'   \item min_sigma, max_sigma - Numeric. Bounds on the subject-level
#'   overdispersion.
#'   \item min_phi, max_phi - Numeric. Bounds on the cell-level overdispersion.
#'   \item cutoff_cell - Numeric. When to refit both overdispersions.
#'   \item kappa - Numeric. When to trust the stage-one subject overdispersion.
#'   \item cpc - Numeric. Minimum mean count per cell for a gene to be tested.
#'   \item mincp - Integer. Minimum number of cells expressing a gene.
#'   \item eps - Numeric. Optimiser stopping tolerance.
#'   \item gene_batch_size - Integer. Genes read and fitted per batch.
#'   \item shrink_dispersion - Boolean. Empirical Bayes shrinkage of the
#'   cell-level overdispersions.
#' }
#' @param .verbose Boolean or integer. Controls verbosity and returns run
#' times. `FALSE` -> quiet, `TRUE` or `1L` -> normal verbosity, `2L` ->
#' detailed verbosity.
#'
#' @returns A `ScNebula` class, see `bixverse:::new_sc_nebula_res()`, with
#' \itemize{
#'   \item results - data.table. One row per gene that survived NEBULA's
#'   expression filter, with the Wald test and both overdispersions.
#'   \item coefficients - Numeric matrix of genes x coefficients.
#'   \item se - Numeric matrix of genes x coefficients.
#'   \item params - List. The parameters the run used.
#' }
#'
#' @export
#'
#' @references He, et al., Commun Biol, 2021
nebula_gpu_sc <- S7::new_generic(
  name = "nebula_gpu_sc",
  dispatch_args = "object",
  fun = function(
    object,
    subject_col,
    design,
    coef = NULL,
    contrast = NULL,
    genes_to_use = NULL,
    offset = NULL,
    nebula_params = params_nebula_gpu(),
    .verbose = TRUE
  ) {
    assert_gpu()

    S7::S7_dispatch()
  }
)

## SingleCells -----------------------------------------------------------------

S7::method(nebula_gpu_sc, SingleCells) <- function(
  object,
  subject_col,
  design,
  coef = NULL,
  contrast = NULL,
  genes_to_use = NULL,
  offset = NULL,
  nebula_params = params_nebula_gpu(),
  .verbose = TRUE
) {
  .nebula_gpu(
    object = object,
    subject_col = subject_col,
    design = design,
    coef = coef,
    contrast = contrast,
    genes_to_use = genes_to_use,
    offset = offset,
    nebula_params = nebula_params,
    .verbose = .verbose
  )
}

## SingleCellsSubset -----------------------------------------------------------

S7::method(nebula_gpu_sc, SingleCellsSubset) <- function(
  object,
  subject_col,
  design,
  coef = NULL,
  contrast = NULL,
  genes_to_use = NULL,
  offset = NULL,
  nebula_params = params_nebula_gpu(),
  .verbose = TRUE
) {
  .nebula_gpu(
    object = object,
    subject_col = subject_col,
    design = design,
    coef = coef,
    contrast = contrast,
    genes_to_use = genes_to_use,
    offset = offset,
    nebula_params = nebula_params,
    .verbose = .verbose
  )
}

## implementation --------------------------------------------------------------

#' GPU NEBULA implementation
#'
#' @description
#' Shared body of the [nebula_gpu_sc()] methods. Mirrors the CPU
#' `bixverse::nebula_sc()` method step for step and only swaps the Rust call.
#'
#' @inheritParams nebula_gpu_sc
#'
#' @returns A `ScNebula` class.
#'
#' @keywords internal
.nebula_gpu <- function(
  object,
  subject_col,
  design,
  coef,
  contrast,
  genes_to_use,
  offset,
  nebula_params,
  .verbose
) {
  # checks
  checkmate::assertTRUE(
    S7::S7_inherits(object, bixverse::SingleCells) ||
      S7::S7_inherits(object, bixverse::SingleCellsSubset)
  )
  checkmate::qassert(subject_col, "S1")
  checkmate::assertFormula(design)
  checkmate::qassert(genes_to_use, c("0", "S+"))
  checkmate::qassert(offset, c("0", "N+"))
  assertNebulaGpuParams(nebula_params)
  checkmate::qassert(.verbose, c("B1", "I1[0,2]"))

  obs <- bixverse::get_sc_obs(
    object,
    cols = unique(c("cell_idx", subject_col, all.vars(design))),
    filtered = TRUE
  )

  inputs <- bixverse:::.nebula_design(
    obs = obs,
    design = design,
    subject_col = subject_col
  )
  tested <- bixverse:::.resolve_tested_sc(
    design = inputs$design_mat,
    coef = coef,
    contrast = contrast
  )

  # obs cell_idx is 1-based, Rust wants 0-based global positions
  cells_to_keep <- as.integer(inputs$obs$cell_idx - 1L)

  if (!is.null(offset)) {
    checkmate::assertNumeric(
      offset,
      len = length(cells_to_keep),
      lower = .Machine$double.eps,
      any.missing = FALSE
    )
  }

  # Rust wants one subject label per cell in the store, not per selected cell.
  # The count comes off the store: a subset shares its parent's counts file and
  # keeps `cell_idx` in the parent's index space.
  n_cells_total <- bixverse::get_sc_rust_ptr(object)$get_shape()[1]
  subject_ids <- integer(n_cells_total)
  subject_ids[inputs$obs$cell_idx] <- as.integer(inputs$subject_fct) - 1L

  gene_indices <- if (is.null(genes_to_use)) {
    seq_along(bixverse::get_gene_names(object)) - 1L
  } else {
    bixverse::get_gene_indices(
      x = object,
      gene_ids = genes_to_use,
      rust_index = TRUE
    )
  }

  res <- rs_nebula_sc_gpu(
    f_path_genes = bixverse:::get_rust_count_gene_f_path(object),
    f_path_cells = bixverse:::get_rust_count_cell_f_path(object),
    cells_to_keep = cells_to_keep,
    gene_indices = as.integer(gene_indices),
    subject_ids = subject_ids,
    design = inputs$design_mat,
    offset = offset,
    nebula_params = c(nebula_params, tested),
    verbose = parse_verbosity(.verbose)
  )

  params <- list(
    subject_col = subject_col,
    design = deparse1(design),
    tested = if (is.null(tested$coef)) {
      "contrast"
    } else {
      colnames(inputs$design_mat)[tested$coef + 1L]
    },
    n_cells = length(cells_to_keep),
    n_subjects = nlevels(inputs$subject_fct),
    n_genes_requested = length(gene_indices),
    nebula_params = nebula_params
  )

  bixverse:::.nebula_res_to_class(
    res = res,
    gene_names = bixverse::get_gene_names(object),
    design_mat = inputs$design_mat,
    params = params
  )
}
